File: tests/algorithms/x/Pascal/bit_manipulation/highest_set_bit.pas
{$mode objfpc}
program Main;
uses SysUtils;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  number: integer;
function highest_set_bit_position(number: integer): integer; forward;
function highest_set_bit_position(number: integer): integer;
var
  highest_set_bit_position_position: integer;
  highest_set_bit_position_n: integer;
begin
  if number < 0 then begin
  panic('number must be non-negative');
end;
  highest_set_bit_position_position := 0;
  highest_set_bit_position_n := number;
  while highest_set_bit_position_n > 0 do begin
  highest_set_bit_position_position := highest_set_bit_position_position + 1;
  highest_set_bit_position_n := highest_set_bit_position_n div 2;
end;
  exit(highest_set_bit_position_position);
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  writeln(IntToStr(highest_set_bit_position(25)));
  writeln(IntToStr(highest_set_bit_position(37)));
  writeln(IntToStr(highest_set_bit_position(1)));
  writeln(IntToStr(highest_set_bit_position(4)));
  writeln(IntToStr(highest_set_bit_position(0)));
  Sleep(1);
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.

File: tests/algorithms/x/Clojure/bit_manipulation/highest_set_bit.clj
(ns main (:refer-clojure :exclude [highest_set_bit_position]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(defn indexOf [s sub]
  (let [idx (clojure.string/index-of s sub)] (if (nil? idx) -1 idx)))

(defn split [s sep]
  (clojure.string/split s (re-pattern sep)))

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare highest_set_bit_position)

(def ^:dynamic highest_set_bit_position_n nil)

(def ^:dynamic highest_set_bit_position_position nil)

(defn highest_set_bit_position [highest_set_bit_position_number]
  (binding [highest_set_bit_position_n nil highest_set_bit_position_position nil] (try (do (when (< highest_set_bit_position_number 0) (throw (Exception. "number must be non-negative"))) (set! highest_set_bit_position_position 0) (set! highest_set_bit_position_n highest_set_bit_position_number) (while (> highest_set_bit_position_n 0) (do (set! highest_set_bit_position_position (+ highest_set_bit_position_position 1)) (set! highest_set_bit_position_n (quot highest_set_bit_position_n 2)))) (throw (ex-info "return" {:v highest_set_bit_position_position}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (println (str (highest_set_bit_position 25)))
      (println (str (highest_set_bit_position 37)))
      (println (str (highest_set_bit_position 1)))
      (println (str (highest_set_bit_position 4)))
      (println (str (highest_set_bit_position 0)))
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)
